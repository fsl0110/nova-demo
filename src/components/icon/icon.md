### Icon name

```jsx
<h5>search</h5>
<Icon name="search" size="xl"/>
<h5>refresh</h5>
<Icon name="refresh" size="xl"/>
<h5>arrow</h5>
<Icon name="arrow" size="xl"/>
<h5>close</h5>
<Icon name="close" size="xl"/>
```

### Icon size

```jsx
<div>
  <h5>size xs</h5>
  <Icon name="refresh" size="xs" />
  <h5>size s</h5>
  <Icon name="refresh" size="s" />
  <h5>size m</h5>
  <Icon name="refresh" size="m" />
  <h5>size l</h5>
  <Icon name="refresh" size="l" />
  <h5>size xl</h5>
  <Icon name="refresh" size="xl" />
</div>
```

### Icon rotate

```jsx
<div>
  <h5>0 degrees</h5>
  <Icon name="search" size="xl" />
  <h5>90 degrees</h5>
  <Icon name="search" rotate="90" size="xl" />
  <h5>180 degrees</h5>
  <Icon name="search" rotate="180" size="xl" />
  <h5>270 degrees</h5>
  <Icon name="search" rotate="270" size="xl" />
</div>
```
