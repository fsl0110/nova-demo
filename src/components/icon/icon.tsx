import React, { PureComponent, HTMLAttributes } from "react";
import classNames from "classnames";
import sprite from "./spritemap.svg";

export type IconSizeProps = "xs" | "s" | "m" | "l" | "xl";

export type IconRotateProps = null | "90" | "180" | "270";

export type IconNameProps = "search" | "refresh" | "arrow" | "close";

export interface Props extends HTMLAttributes<HTMLOrSVGElement> {
  size: IconSizeProps;
  /** This component does not support custom children. Use the `icon` prop. */
  children?: never;
  /** Name of the icon to show. */
  name?: IconNameProps;
  /** Rotation of the icon e.g. to use one Arrow-Icon for all directions. */
  rotate?: IconRotateProps;
  /** Spin of the icon. If set to true the icon works as a spinner. */
  spin?: boolean;
}

export class Icon extends PureComponent<Props> {
  static defaultProps: Partial<Props> = {
    size: "s"
  };

  render() {
    const { name, rotate, size, spin, className, ...rest } = this.props;

    if (!name) {
      return null;
    }

    const rootClasses = classNames("fs-icon", className, {
      [`fs-icon--${size}`]: !!size,
      [`fs-icon--${rotate}`]: !!rotate,
      "fs-icon--spin": spin
    });

    return (
      <svg
        role="presentation"
        focusable="false"
        className={rootClasses}
        data-test="Icon"
        {...rest}
      >
        {<use xlinkHref={`${sprite}#sprite-${name}`} />}
      </svg>
    );
  }
}

export default Icon;
