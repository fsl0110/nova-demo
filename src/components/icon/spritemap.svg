<svg xmlns="http://www.w3.org/2000/svg"><symbol id="sprite-arrow" viewBox="0 0 24 24"><title>arrow</title><path d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"/><path d="M0 0h24v24H0z" fill="none"/></symbol><symbol id="sprite-close" viewBox="0 0 24 24"><title>close</title><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/><path d="M0 0h24v24H0z" fill="none"/></symbol><symbol id="sprite-refresh" viewBox="0 0 24 24"><title>refresh</title><path d="M17.65 6.35A7.958 7.958 0 0012 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08A5.99 5.99 0 0112 18c-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"/><path d="M0 0h24v24H0z" fill="none"/></symbol><symbol id="sprite-search" viewBox="0 0 24 24"><title>search</title><path d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"/><path d="M0 0h24v24H0z" fill="none"/></symbol></svg>