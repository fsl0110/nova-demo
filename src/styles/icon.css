.fs-icon {
  display: inline-block;
  pointer-events: none;
  width: inherit;
  height: inherit;
  fill: currentColor;
  color: inherit;
  z-index: 1;
}

.fs-icon--spin {
  animation: spin 0.6s linear infinite;
}

@keyframes spin {
  to {
    transform: rotate(360deg);
  }
}

.fs-icon--90 {
  transform: rotate(90deg);
}

.fs-icon--180 {
  transform: rotate(90deg);
}

.fs-icon--270 {
  transform: rotate(90deg);
}

.fs-icon--xs {
  width: 0.8em;
  height: 0.8em;
}

.fs-icon--s {
  width: 1em;
  height: 1em;
}

.fs-icon--m {
  width: 1.4em;
  height: 1.4em;
}

.fs-icon--l {
  width: 2em;
  height: 2em;
}

.fs-icon--xl {
  width: 2.8em;
  height: 2.8em;
}
