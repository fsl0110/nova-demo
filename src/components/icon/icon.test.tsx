import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Icon, Props } from "./icon";

const testCases: Array<[string, Props]> = [
  ["Icon returns null", {}],
  [
    "Icon refresh default",
    {
      name: "refresh"
    }
  ],
  [
    "Icon refresh size xl",
    {
      name: "refresh",
      size: "xl"
    }
  ],
  [
    "Icon search rotation 180",
    {
      name: "search",
      rotate: "180"
    }
  ],
  [
    "Icon search rotation 180 size xl",
    {
      name: "search",
      rotate: "180",
      size: "xl"
    }
  ]
];

describe("<Icon>", () => {
  testCases.forEach(([item, options]) => {
    it(item, () => {
      expect(toJson(shallow(<Icon {...options} />))).toMatchSnapshot();
    });
  });
});
