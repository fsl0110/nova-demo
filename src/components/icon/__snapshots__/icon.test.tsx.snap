// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon> Icon refresh default 1`] = `
<svg
  className="fs-icon fs-icon--s"
  data-test="Icon"
  focusable="false"
  role="presentation"
>
  <use
    xlinkHref="spritemap.svg#sprite-refresh"
  />
</svg>
`;

exports[`<Icon> Icon refresh size xl 1`] = `
<svg
  className="fs-icon fs-icon--xl"
  data-test="Icon"
  focusable="false"
  role="presentation"
>
  <use
    xlinkHref="spritemap.svg#sprite-refresh"
  />
</svg>
`;

exports[`<Icon> Icon returns null 1`] = `""`;

exports[`<Icon> Icon search rotation 180 1`] = `
<svg
  className="fs-icon fs-icon--s fs-icon--180"
  data-test="Icon"
  focusable="false"
  role="presentation"
>
  <use
    xlinkHref="spritemap.svg#sprite-search"
  />
</svg>
`;

exports[`<Icon> Icon search rotation 180 size xl 1`] = `
<svg
  className="fs-icon fs-icon--xl fs-icon--180"
  data-test="Icon"
  focusable="false"
  role="presentation"
>
  <use
    xlinkHref="spritemap.svg#sprite-search"
  />
</svg>
`;
